include/gabia/secure_socket.hpp

```diff
@@ -0,0 +1,302 @@
+            // TODO: use a more appropriate error code
+            error = asio::error::no_permission;
+            decrypted_buffer.consume(payload_length);
+            return 0;
+        }
+
+        auto consumed =
+            boost::asio::buffer_copy(sequence, decrypted_buffer.data());
+        decrypted_buffer.consume(consumed);
+        return consumed;
+    }
+
+    template <typename MutableBufferSequence>
+    std::size_t read_some(const MutableBufferSequence& sequence) {
+        beast::error_code error = {};
+        auto n = read_some(sequence, error);
+        Expects(!error);
+        return n;
+    }
+
+    template <typename MutableBufferSequence, typename Handler>
+    auto async_read_some(MutableBufferSequence sequence, Handler&& handler) {
+        beast::async_completion<Handler, void(beast::error_code, size_t)> init{
+            handler};
+        if (decrypted_buffer.size() > 0) {
+            auto consumed =
+                boost::asio::buffer_copy(sequence, decrypted_buffer.data());
+            decrypted_buffer.consume(consumed);
+            next_layer().get_io_service().dispatch([
+                consumed, handler = std::move(init.completion_handler)
+            ]() mutable {
+                beast::error_code error{};
+                handler(error, consumed);
+            });
+        } else {
+            async_read_some_op<MutableBufferSequence, Handler> operation{
+                *this, sequence, std::move(init.completion_handler)};
+            next_layer().get_io_service().dispatch([operation = std::move(
+                                                        operation)]() mutable {
+                beast::error_code error{};
+                operation(error, 0);
+            });
+        }
+        return init.result.get();
+    }
+
+    template <typename ConstBufferSequence, typename Handler>
+    auto async_write_some(ConstBufferSequence sequence, Handler&& handler) {
+        beast::async_completion<Handler, void(beast::error_code, size_t)> init{
+            handler};
+        async_write_some_op<ConstBufferSequence, Handler> operation{
+            *this, sequence, std::move(init.completion_handler)};
+        next_layer().get_io_service().dispatch([operation = std::move(
+                                                    operation)]() mutable {
+            beast::error_code error{};
+            operation(error, 0);
+        });
+        return init.result.get();
+    }
+
+    void set_secrets(const crypto::aead_secrets& secrets) {
+        this->secrets = secrets;
+    }
+
+private:
+    template <typename ConstBufferSequence, typename Handler>
+    struct async_write_some_op : public asio::coroutine {
+        async_write_some_op(secure_socket& socket,
+                            const ConstBufferSequence& sequence,
+                            Handler&& handler)
+            : socket{socket},
+              sequence{sequence},
+              handler{std::forward<Handler>(handler)} {}
```

Comment:
This isn't going to work the way you think. You are not in a "deduced context", because `Handler` is a class template parameter not a function template parameter. Your choice is either to have two signatures for the constructor (`Handler const&` and `Handler&&`) or you can do what I do which is just make the constructor templated:
```
template<class DeducedHandler>
async_write_some_op(
    secure_socket& socket,
    const ConstBufferSequence& sequence,
    DeducedHandler&& handler)
    : handler_(std::forward<DeducedHandler>(handler)
{
}
```

I notice you have the same name `handler` as both a data member and a function argument, that is highly confusing :( and I'm not sure its defined behavior.